appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/interceptors/CallbackInvocationContext.java

```diff
@@ -227,42 +225,43 @@ public void invokeSpecial() throws Throwable {
     }
 
     private void checkSetParameters(Object[] params) {
-       if( ctor != null) {
-
-            if ((params == null) && (ctorParamTypes.length != 0)) {
-                throw new IllegalArgumentException("Wrong number of parameters for "
-                        + " constructor: " + ctor);
-            }
-            if (ctorParamTypes.length != (params != null ? params.length : 0)) {
-                throw new IllegalArgumentException("Wrong number of parameters for "
-                        + " constructor: " + ctor);
-            }
-            int index = 0 ;
-            for (Class type : ctorParamTypes) {
-                if (params[index] == null) {
-                    if (type.isPrimitive()) {
-                        throw new IllegalArgumentException("Parameter type mismatch for constructor "
-                                + ctor + ".  Attempt to set a null value for Arg["
-                            + index + "]. Expected a value of type: " + type.getName());
-                    }
-                } else if (type.isPrimitive()) {
-                    if (! InterceptorUtil.hasCompatiblePrimitiveWrapper(type, params[index].getClass())) {
-                        throw new IllegalArgumentException("Parameter type mismatch for constructor "
-                                + ctor + ".  Arg["
-                            + index + "] type: " + params[index].getClass().getName()
-                            + " is not compatible with the expected type: " + type.getName());
-                    }
-                } else if (! type.isAssignableFrom(params[index].getClass())) {
-                    throw new IllegalArgumentException("Parameter type mismatch for constructor "
-                            + ctor + ".  Arg["
-                        + index + "] type: " + params[index].getClass().getName()
-                        + " does not match the expected type: " + type.getName());
-                }
-                index++;
-            }
-        } else {
-            throw new IllegalStateException("Internal Error: Got null constructor");
-        }
+       if( ctor == null) {
+           throw new IllegalStateException("Internal Error: Got null constructor");
+       }
+       if (params == null) {
```

Comment:
`params` needed proper null handling. All cases were practically present just not nested and formulated in a way that makes sure the intended `IllegalArgumentException` is thrown. Moved the `else` from end of method to top to get rid of 1 nesting level for rest of the method body. Behaviour change: NPE becomes IllegalArgumentException (as intended)